### PR TITLE
Remove SharedTree legacy transaction APIs

### PR DESCRIPTION
**Pull request by noencke**
[How contribute to this repo](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md).

[Guidelines for Pull Requests](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

## Description

This updates the transaction APIs to be more cohesive. It replaces the use of the old `Checkout` (or `TransactionCheckout`, as it was aliased in practice) with the new transaction API on `ISharedTreeCheckout`. It provides a convenience helper `runSynchronous` to help with the common case of a simple, short-lived transaction. It moves `runSynchronousTransaction` (renamed to `runTransactionOnForest` to be a test utility, as it is still useful for `TestTree` and forest unit tests.

## Breaking Changes

* This removes `runTransaction` on ISharedTreeCheckout. The equivalent replacement is the free function `runSynchronous`.

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts [hunk 1]
```diff
@@ -308,13 +305,16 @@ export class ProxyContext implements EditableTreeContext {
...
- * @param transactionCheckout - the TransactionCheckout applied to a transaction, not required in read-only usecases.
+ * @param runTransaction - a function which executes a given transaction, returning true if it was successful or false if aborted. Not required in read-only use-cases.
  * @returns {@link EditableTreeContext} which is used to manage the cursors and anchors within the EditableTrees:
  * This is necessary for supporting using this tree across edits to the forest, and not leaking memory.
  */
 export function getEditableTreeContext(
 	forest: IEditableForest,
-	transactionCheckout?: TransactionCheckout<DefaultEditBuilder, DefaultChangeset>,
+	runTransaction?: (
+		/** A function which performs a transaction, returning true if the transaction should succeed or false if it should be aborted */
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:315]: this isn't a good way to document this. Its redundant with the param doc above and the other copy of this comment further above.

Either rely on the param doc, and/or define a type for it and but the docs on that type.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts [hunk 2]
```diff
@@ -130,14 +126,14 @@ export class ProxyContext implements EditableTreeContext {
...
-	 * @param transactionCheckout - the Checkout applied to a transaction, not required in read-only usecases.
+	 * @param runTransaction - a function which executes a given transaction, returning true if it was successful or false if aborted. Not required in read-only use-cases.
 	 */
 	constructor(
 		public readonly forest: IEditableForest,
-		private readonly transactionCheckout?: TransactionCheckout<
-			DefaultEditBuilder,
-			DefaultChangeset
-		>,
+		private readonly runTransaction?: (
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:133]: I don't think this code needs a way to roll back, or do grouping. I think just passing in the `editor` would be more inline with what this code actually wants to do. Its use of transactions was just an artifact of how the olds API forced it to work

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:315]: Note that if you pass in just the editor as suggested above, this goes away

File: packages/dds/tree/src/test/shared-tree/opSize.bench.ts [hunk 3]
```diff
@@ -70,12 +69,11 @@ const childrenFieldKey: FieldKey = brand("children");
...
-		const writeCursor = singleTextCursor(state);
-		const field = editor.sequenceField(undefined, rootFieldKeySymbol);
-		field.insert(0, writeCursor);
-		return TransactionResult.Apply;
-	});
+	tree.transaction.start();
+	const writeCursor = singleTextCursor(state);
+	const field = tree.editor.sequenceField(undefined, rootFieldKeySymbol);
+	field.insert(0, writeCursor);
+	tree.transaction.commit();
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/shared-tree/opSize.bench.ts:76]: this whole edit is a single insert, so it seems unneeded to wrap it in a transaction, right?

File: packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts [hunk 4]
```diff
@@ -139,7 +138,7 @@ export async function performFuzzActionsAbort(
 		{
 			edit: async (state, operation) => {
 				const { index, contents } = operation;
-				applyFuzzChange(tree, contents, TransactionResult.Abort);
+				applyFuzzChange(tree, contents, false);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts:141]: this one used to Abort, and the one above used to Apply.

Now both "false".

Enums are much better than boolean when its not super clear what they mean. When I'm reading this code, I can't what that boolean does. Enums are more tpye safe, and more readable.

I think it makes sense to reintroduce this enum somewhere and use it in places like this for improved clarity.

Comment (noencke) [reply packages/dds/tree/src/test/shared-tree/sharedTreeFuzzTests.ts:141]: Thanks, copy paste error - good catch. Yea, I'll see what I can do.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts (hunk 2 above)
Comment (noencke) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:133]: Great, that will be much cleaner.

File: packages/dds/tree/src/test/shared-tree/opSize.bench.ts (hunk 3 above)
Comment (noencke) [reply packages/dds/tree/src/test/shared-tree/opSize.bench.ts:76]: Agreed, thanks.

File: packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts [hunk 5]
```diff
@@ -219,55 +212,67 @@ export class ProxyContext implements EditableTreeContext {
...
 		newContent: ITreeCursor,
-	): boolean {
-		return this.runTransaction((editor) => {
-			const field = editor.valueField(path, fieldKey);
-			field.set(newContent);
-		});
+	): void {
+		assert(
+			this.editor !== undefined,
+			0x45a /* an editor is required to edit the EditableTree */,
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:230]: you can not have the same assert ID in multiple locations. That will fail policy check. THis number is suppose to identify a unique location in the codebase. You can use the string message in all these places and let the tagger tool generate numbers for the next time we run it, or avoid having all these checks in the first place.

I think the only place that currently doesn't provide an editor is some tests. We can make those work by providing a mock implementation that errors if you call any of its editing functions (or just cast {} to editor or something). I think this would be cleaner.

Another approach would be to have a getter for editor with the assert in it.

Comment (noencke) [reply packages/dds/tree/src/feature-libraries/editable-tree/editableTreeContext.ts:230]: Oh. I always assumed that if the string matched, then the error number would match. But I guess it makes sense to have duplicates if the point is to identify a specific line of code.

I like the mock idea too, I'll do that.

File: packages/dds/tree/src/test/utils.ts [hunk 6]
```diff
@@ -423,13 +430,72 @@ export function initializeTestTree(
...
+
+	if (aborted) {
+		return TransactionResult.Abort;
+	}
+
+	// Using anonymous changes makes it impossible to chronologically order them during composition.
+	// Such an ordering is needed when composing/squashing inverse changes with other changes, which is currently
+	// not expected to happen in transactions but that could change in the future.
+	const anonChanges = changes.map((c) => makeAnonChange(c));
+	submitEdit(changeFamily.rebaser.compose(anonChanges));
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/utils.ts:499]: What is this edit being submitted to?

File: packages/dds/tree/src/test/utils.ts [hunk 7]
```diff
@@ -423,13 +430,72 @@ export function initializeTestTree(
...
+	// then reapply it (when the local edit is added) when possible.
+	{
+		// Roll back changes
+		for (const inverse of inverses) {
+			// TODO: maybe unify logic to edit forest and its anchors here with that in ProgressiveEditBuilder.
+			// TODO: update schema in addition to anchors and tree data (in both places).
+			changeFamily.rebaser.rebaseAnchors(forest.anchors, inverse);
+			forest.applyDelta(changeFamily.intoDelta(inverse));
+		}
+	}
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/utils.ts:489]: isn't the point of this to apply the changes to the forest? Why unapply them?

Comment (CraigMacomber) [reply packages/dds/tree/src/test/utils.ts:489]: THis seems like it should be in the abort only case

File: packages/dds/tree/src/test/utils.ts [hunk 8]
```diff
@@ -423,13 +430,72 @@ export function initializeTestTree(
...
+ * @param changeFamily - the change family of the edits
+ * @param submitEdit - a callback that provides the change whenever the forest is mutated
+ * @param transaction - the transaction to run. Use the editor passed to this function to mutate the forest.
+ * If it returns an `Abort` result then the transaction will be aborted. Otherwise, it will be committed.
+ * @returns true if the transaction was committed, otherwise false
+ */
+export function runTransactionOnForest<TEditor extends ProgressiveEditBuilder<TChange>, TChange>(
+	forest: IEditableForest,
+	changeFamily: ChangeFamily<TEditor, TChange>,
+	submitEdit: (edit: TChange) => void,
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/utils.ts:451]: it seems like either runTransactionOnForest should output an edit, ot it should update the forest (and anchotSet) to the in the modified state. It is confusing to have this submitEdit callback here.

Maybe document that this function leaves the forest in its inital state, and returns a TChange (then make the TChange the return value)?

Comment (CraigMacomber) [reply packages/dds/tree/src/test/utils.ts:451]: really seems like we should just update our tests to do things the new way, instead of making the old codepath a test helper

File: packages/dds/tree/src/shared-tree/sharedTree.ts [hunk 9]
```diff
@@ -123,28 +122,17 @@ export interface ISharedTreeCheckout extends AnchorLocator {
 		 * Close this transaction by squashing its edits and committing them as a single edit. If this is the root local branch and there are no
 		 * ongoing transactions remaining, the squashed edit will be submitted to Fluid.
 		 */
-		commit: () => void;
+		commit: () => TransactionResult.Commit;
```
Comment (alex-pardes) [inline packages/dds/tree/src/shared-tree/sharedTree.ts:125]: What is the purpose of this change?

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts [hunk 10]
```diff
@@ -53,79 +48,58 @@ describe("ChunkedForest", () => {
 	});
 
 	it("doesn't copy data when capturing and restoring repair data", () => {
```
Comment (alex-pardes) [inline packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:50]: How are the changes to this test related to the rest of this PR?

File: packages/dds/tree/src/util/transactionResult.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Describes the result of a transaction. Transactions may either succeed and commit, or fail and abort.
+ * @alpha
+ */
+export enum TransactionResult {
+	/**
+	 * Indicates the transaction succeeded. This value is falsy.
```
Comment (alex-pardes) [inline packages/dds/tree/src/util/transactionResult.ts:12]: What is the motivation for making `Commit` falsy and `Abort` truthy?

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/util/transactionResult.ts (hunk 11 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/util/transactionResult.ts:12]: No idea.

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 10 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:50]: This test had a lot of needless dependencies (ex: doing most of the end to end editing process), which made it (and another test I fixed) the only users of a bunch of helper logic to replicate the old code thats deleted. These changes reduce its dependencies and simplify whats done here.

File: api-report/tree.api.md [hunk 12]
```diff
@@ -1168,12 +1170,10 @@ export interface TaggedChange<TChangeset> {
 // @alpha
 export type ToDelta = (child: NodeChangeset) => Delta.Modify;
 
-// @alpha (undocumented)
+// @alpha
 export enum TransactionResult {
```
Comment (yann-achard-MS) [inline api-report/tree.api.md:1174]: The PR description needs updating to reflect these changes.

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts [hunk 13]
```diff
@@ -53,79 +48,58 @@ describe("ChunkedForest", () => {
...
-			fields: {
-				foo: [
-					{ type: brand("Number"), value: 0 },
-					{ type: brand("Number"), value: 1 },
-					{ type: brand("Number"), value: 2 },
-				],
-			},
-		};
-		const anchors = new AnchorSet();
+		const initialState: JsonableTree = { type: brand("Node") };
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:51]: When authoring tests with tree data, Jenn and I discovered recently that it's a often good thing to use a non-trivial tree (i.e., a tree with a populated field). Without the field, conversion bugs between our tree represenations (which our typing doesn't always detect) can remain undetected.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:51]: We probably don't need three child nodes though.

File: packages/dds/tree/src/shared-tree/sharedTree.ts (hunk 9 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/shared-tree/sharedTree.ts:125]: I assume its for ergononomics, so you can do `return commit();` or `return abort();` in some cases. That said the cases that could leverage this are not, I'll update them

Also I don't like making these all writable members instead of readonly function or methods! I'll fix that too.

Comment (CraigMacomber) [reply packages/dds/tree/src/shared-tree/sharedTree.ts]: Done.

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 13 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:51]: This is a unit test for a very specific thing. Having tests which don't get broken by unrelated bugs is good.
Any tests that larger trees work properly should not be here as we wouldn't want such a test to be specific to this one forest implementation.

Also this test is ensuring that the underlying tree does not go through any conversion (reused the same exact object instance). It seems unlikely that a conversion could break children if the same object is uses and thus the children must remain exactly as they were.

File: api-report/tree.api.md (hunk 12 above)
Comment (CraigMacomber) [reply api-report/tree.api.md:1174]: done.

File: packages/dds/tree/src/util/transactionResult.ts (hunk 11 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/util/transactionResult.ts]: I have removed this. It can't even be used within our package with our current linter rules (as of today).

File: packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 13 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:51]: I do think the general guidance of at least have a test for larger data cases is good though. I just don't think this is the place to add such a test. Also I think having minimal tests and larger cases +fuzz testing is a batter approach than only having a larger test. Minimal tests are great for bug localization, debugging, and as simple examples of what the feature is, and I don't want to lose that.

File: packages/dds/tree/src/test/shared-tree/testTree.ts [hunk 14]
```diff
@@ -139,34 +126,36 @@ export class TestTree {
...
+
+		// Using anonymous changes makes it impossible to chronologically order them during composition.
+		// Such an ordering is needed when composing/squashing inverse changes with other changes, which is currently
+		// not expected to happen in transactions but that could change in the future.
+		const anonChanges = changes.map((c) => makeAnonChange(c));
+
+		const changeset = defaultChangeFamily.rebaser.compose(anonChanges);
 		const revision = mintRevisionTag();
-		const delta = this.editManager.addLocalChange(revision, changeset);
+		// Forest and anchors already reflect the results of this change since they were updated during transaction.
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/test/shared-tree/testTree.ts:150]: These changes are problematic in that they introduce a divergence between how SharedTree transactions work and how TestTree transactions work.

We need to have a conversation about how to avoid having this code in the first place, meaning that the tests should leverage the production code, but I' rather not make things worse in the meantime.

It should be possible to update this code so that it still mimics what happens in the production code.

Comment (CraigMacomber) [reply packages/dds/tree/src/test/shared-tree/testTree.ts:150]: Before this PR this test this code would apply the changes, roll them back, then submit+reapply them using a bunch of test only code. My understanding is SharedTree doesn't do this for its editing APIs anymore? This PR makes this test no longer take that approach.

Comment (yann-achard-MS) [reply packages/dds/tree/src/test/shared-tree/testTree.ts:150]: You're right. I had no idea!

**Review by yann-achard-MS — APPROVED** [review]
